crates/store/re_video/src/decode/av1.rs

```diff
@@ -0,0 +1,408 @@
+use dav1d::{PixelLayout, PlanarImageComponent};
+
+use crate::Time;
+
+use super::{Chunk, Frame, PixelFormat};
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    /// An error occrurred during initialization of the decoder.
+    ///
+    /// No further output will come.
+    #[error("Error initializing decoder: {0}")]
+    Initialization(dav1d::Error),
+
+    #[error("Decoding error: {0}")]
+    Dav1d(dav1d::Error),
+}
+
+pub type Result<T = (), E = Error> = std::result::Result<T, E>;
+
+enum Command {
+    Chunk(Chunk),
+    Flush { on_done: Sender<()> },
+    Reset,
+    Stop,
+}
+
+#[derive(Clone)]
+struct Comms {
+    /// Set when it is time to die
+    should_stop: Arc<AtomicBool>,
+
+    /// Incremented on each call to [`Decoder::reset`].
+    /// Decremented each time the decoder thread receives [`Command::Reset`].
+    num_outstanding_resets: Arc<AtomicU64>,
+}
+
+impl Default for Comms {
+    fn default() -> Self {
+        Self {
+            should_stop: Arc::new(AtomicBool::new(false)),
+            num_outstanding_resets: Arc::new(AtomicU64::new(0)),
+        }
+    }
+}
+
+/// AV1 software decoder.
+pub struct Decoder {
+    /// Where the decoding happens
+    _thread: std::thread::JoinHandle<()>,
+
+    /// Commands sent to the decoder thread.
+    command_tx: Sender<Command>,
+
+    /// Instant communication to the decoder thread (circumventing the command queue).
+    comms: Comms,
+}
+
+impl Decoder {
+    pub fn new(
+        debug_name: String,
+        on_output: impl Fn(Result<Frame>) + Send + Sync + 'static,
+    ) -> Self {
+        re_tracing::profile_function!();
+        let (command_tx, command_rx) = unbounded();
+        let comms = Comms::default();
+
+        let thread = std::thread::Builder::new()
+            .name("av1_decoder".into())
+            .spawn({
+                let comms = comms.clone();
+                move || {
+                    econtext::econtext_data!("Video", debug_name.clone());
+                    decoder_thread(&comms, &command_rx, &on_output);
+                    re_log::debug!("Closing decoder thread for {debug_name}");
+                }
+            })
+            .expect("failed to spawn decoder thread");
+
+        Self {
+            _thread: thread,
+            command_tx,
+            comms,
+        }
+    }
+
+    // NOTE: The interface is all `&mut self` to avoid certain types of races.
+    pub fn decode(&mut self, chunk: Chunk) {
+        re_tracing::profile_function!();
+        self.command_tx.send(Command::Chunk(chunk)).ok();
+    }
+
+    /// Resets the decoder.
+    ///
+    /// This does not block, all chunks sent to `decode` before this point will be discarded.
+    // NOTE: The interface is all `&mut self` to avoid certain types of races.
+    pub fn reset(&mut self) {
+        re_tracing::profile_function!();
+        self.comms
+            .num_outstanding_resets
+            .fetch_add(1, Ordering::SeqCst);
```

Comment:
SeqCst always makes me nervous because it's so hard to reason with. This should be `Release` because we want to ensure that at least all the writes so far are visible to the readers (this may or may not include future writes), i.e. at least all previous operations on the channel are published.
On that note, I think an explanation on why it **has** to be before the `send` would be helpful.